ctapipe/io/tableloader.py

```diff
@@ -0,0 +1,330 @@
+
+        try:
+            self.structure = get_structure(self.h5file)
+        except ValueError:
+            self.structure = None
+
+        self.trigger_table = read_table(self.h5file, TRIGGER_TABLE)
+
+        self.instrument_table = None
+        if self.load_instrument:
+            table = self.subarray.to_table()
+            optics = self.subarray.to_table(kind="optics")
+            optics["optics_index"] = np.arange(len(optics))
+            optics.remove_columns(["name", "description", "type"])
+            table = join(
+                table,
+                optics,
+                keys="optics_index",
+                # conflicts for TAB_VER, TAB_TYPE, not needed here, ignore
+                metadata_conflicts="silent",
+            )
+
+            table.remove_columns(["optics_index", "camera_index"])
+            self.instrument_table = table
+
+        self.shower_table = None
+        if self.load_simulated and SHOWER_TABLE in self.h5file:
+            self.shower_table = read_table(self.h5file, SHOWER_TABLE)
+
+    def close(self):
+        self.h5file.close()
+
+    def __del__(self):
+        self.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
+
+    def _get_reconstructors(self):
+
+        if GEOMETRY_GROUP in self.h5file:
+            g = self.h5file.root[GEOMETRY_GROUP]
+        else:
+            raise ValueError(f"No DL2 geometry data in h5file: {self.h5file}")
+        
+        reconstructors = g._v_children.keys()
+        
+        return reconstructors
+
+    def _read_telescope_table(self, group, tel_id):
+
+        if self.structure == "by_id":
+            key = f"{group}/tel_{tel_id:03d}"
+            condition = None
+        else:
+            key = f"{group}/{self.subarray.tel[tel_id]!s}"
+            condition = f"tel_id == {tel_id}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=condition)
+        else:
+            table = Table()
+
+        return table
+    
+    def _read_telescope_DL2_geometry(self, group, reconstructor):
+        
+        key = f"{group}/{reconstructor}"
+
+        if key in self.h5file:
+            table = read_table(self.h5file, key, condition=None)
+        else:
+            table = Table()
+
+        return table
+
+    def read_subarray_events(self, labels=None, sort_by=["obs_id", "event_id"]):
+        """Read data from file.
+        
+        Parameters
+        ----------
+        labels: list
+            Either tel_ids or tel_types depending on dataset split.
+        sort_by: list
+            Columns to sort the final table.
+            By default it is sorted by obs_id and event_id.
+        
+        Returns
+        -------
+        output: astropy.Table
+            Single table containing the enabled data information."""
+        
+        if not any(self.all_loads):
+            self.close()
```

Comment:
Don't close! This will prevent people from changing the options and runnig read_subarray_events again.

Close will by called when using this class in a context or when it goes out of scope automatically. There is no need to use close manually anywhere else

Reply:
removes